packages/xsnap/src/xsnap.c

```diff
@@ -0,0 +1,1350 @@
+		if (argv[argi][0] != '-')
+			continue;
+		if (!strcmp(argv[argi], "-f")) {
+			freeze = 1;
+		}
+		else if (!strcmp(argv[argi], "-h")) {
+			fxPrintUsage();
+			return 0;
+		} else if (!strcmp(argv[argi], "-i")) {
+			argi++;
+			if (argi < argc)
+				interval = atoi(argv[argi]);
+			else {
+				fxPrintUsage();
+				return 1;
+			}
+		}
+		else if (!strcmp(argv[argi], "-l")) {
+			argi++;
+			if (argi < argc)
+				gxMeteringLimit = atoi(argv[argi]);
+			else {
+				fxPrintUsage();
+				return 1;
+			}
+		}
+		else if (!strcmp(argv[argi], "-p"))
+			gxMeteringPrint = 1;
+		else if (!strcmp(argv[argi], "-r")) {
+			argi++;
+			if (argi < argc)
+				argr = argi;
+			else {
+				fxPrintUsage();
+				return 1;
+			}
+		}
+		else if (!strcmp(argv[argi], "-v")) {
+			printf("XS %d.%d.%d\n", XS_MAJOR_VERSION, XS_MINOR_VERSION, XS_PATCH_VERSION);
+			return 0;
+		} else {
+			fxPrintUsage();
+			return 1;
+		}
+	}
+	if (gxMeteringLimit) {
+		if (interval == 0)
+			interval = 1;
+	}
+	fxInitializeSharedCluster();
+	if (argr) {
+		snapshot.stream = fopen(argv[argr], "rb");
+		if (snapshot.stream) {
+			machine = fxReadSnapshot(&snapshot, "xsnap", NULL);
+			fclose(snapshot.stream);
+		}
+		else
+			snapshot.error = errno;
+		if (snapshot.error) {
+			fprintf(stderr, "cannot read snapshot %s: %s\n", argv[argr], strerror(snapshot.error));
+			return 1;
+		}
+	}
+	else {
+		machine = xsCreateMachine(creation, "xsnap", NULL);
+		fxBuildAgent(machine);
+		fxPatchBuiltIns(machine);
+	}
+	if (freeze) {
+		fxFreezeBuiltIns(machine);
+		fxShareMachine(machine);
+		fxCheckAliases(machine);
+		machine = xsCloneMachine(creation, machine, "xsnap", NULL);
+	}
+	if (!(fromParent = fdopen(3, "rb"))) {
+		fprintf(stderr, "fdopen(3) from parent failed\n");
+		c_exit(1);
+	}
+	if (!(toParent = fdopen(4, "wb"))) {
+		fprintf(stderr, "fdopen(4) to parent failed\n");
+		c_exit(1);
+	}
+	xsBeginMetering(machine, fxMeteringCallback, interval);
+	{
+		char done = 0;
+		while (!done) {
+			char* nsbuf;
+			size_t nslen;
+			int readError = fxReadNetString(fromParent, &nsbuf, &nslen);
+			if (readError != 0) {
+				if (feof(fromParent)) {
+					break;
+				} else {
+					fprintf(stderr, "%s\n", fxReadNetStringError(readError));
+					c_exit(1);
+				}
+			}
+			char command = *nsbuf;
+			// fprintf(stderr, "command: len %d %c arg: %s\n", nslen, command, nsbuf + 1);
+			switch(command) {
+			case '?':
+			case 'e':
+				error = 0;
+				char* response = NULL;
+				size_t responseLength = 0;
+				xsBeginHost(machine);
+				{
+					xsVars(3);
+					xsVar(1) = xsArrayBuffer(nsbuf + 1, nslen - 1);
+					xsTry {
+						if (command == '?') {
+							xsVar(2) = xsCall1(xsGlobal, xsID("answerSysCall"), xsVar(1));
+							if (xsTypeOf(xsVar(2)) != xsUndefinedType) {
+								responseLength = fxGetArrayBufferLength(machine, &xsVar(2));
+								response = malloc(responseLength);
+								fxGetArrayBufferData(machine, &xsVar(2), 0, response, responseLength);
```

Comment:
This dispatches `answerSysCall` instead of `onMessage`. The function may return undefined or an `ArrayBuffer`. The array buffer now gets carried with the OK response, after the `.` control code.

Reply:
In what way is this a syscall? It's a delivery, in the terminology of the [vat-worker](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/docs/vat-worker.md) protocol.

Reply:
I’m open to better names. I arrived at this pair of names while writing `README.md`, attempting to explain parent-to-child and child-to-parent “calls” and noticing that one is a reflection of the other. They both accept and return bytes. The parent is always async. The child is always sync.

The fixed point was `sysCall`. From that it seemed sensible that the parent’s `xsnap` maker would accept an optional `answerSysCall` parameter. That established a naming precedent, so when Brian asked for a better name than `onMessage`, we had one ready-made. From there, it just seemed sensible that the `xsnap` process object should have a matching `sysCall` method.

Reply:
Per conversation with @warner and @dckc, pushed a commit to the end that renames syscalls at this layer to commands (verbs handle and issue) to avoid confusion with syscalls at a higher layer of the architecture.